components/datetime/src/skeleton.rs

```diff
@@ -0,0 +1,408 @@
+        write!(formatter, "Expected to find a valid skeleton.")
+    }
+
+    /// A skeleton serialized into a string follows UTS 35.
+    /// https://unicode.org/reports/tr35/tr35-dates.html#Date_Field_Symbol_Table
+    /// This string consists of a symbol that is repeated N times. This string is
+    /// deserialized here into the Skeleton format which is used in memory
+    /// when working with formatting date times.
+    fn visit_str<E>(self, skeleton_string: &str) -> Result<Self::Value, E>
+    where
+        E: de::Error,
+    {
+        Skeleton::try_from(skeleton_string).map_err(|err| {
+            de::Error::invalid_value(
+                de::Unexpected::Other(&format!("{:?} {}", skeleton_string, err)),
+                &"a UTS 35 sorted field symbols representing a skeleton",
+            )
+        })
+    }
+}
+
+#[cfg(feature = "provider_serde")]
+impl<'de> Deserialize<'de> for Skeleton {
+    fn deserialize<D>(deserializer: D) -> Result<Skeleton, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        deserializer.deserialize_str(SkeletonFieldsDeserializeVisitor)
+    }
+}
+
+#[cfg(feature = "provider_serde")]
+impl Serialize for Skeleton {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: ser::Serializer,
+    {
+        let mut string = String::new();
+
+        for field in self.0.iter() {
+            let ch: char = field.symbol.into();
+            for _ in 0..field.length as usize {
+                string.push(ch);
+            }
+        }
+
+        serializer.serialize_str(&string)
+    }
+}
+
+/// Parse a string into a list of fields. This trait implementation validates the input string to
+/// verify that fields are correct. If the fields are out of order, this returns an error that
+/// contains the fields, which gives the callee a chance to sort the fields with the
+/// `From<SmallVec<[fields::Field; 5]>> for Skeleton` trait.
+impl TryFrom<&str> for Skeleton {
+    type Error = SkeletonError;
+    fn try_from(skeleton_string: &str) -> Result<Self, Self::Error> {
+        let mut fields: SmallVec<[fields::Field; 5]> = SmallVec::new();
+        let mut in_order = true;
+
+        let mut iter = skeleton_string.bytes().peekable();
+        while let Some(byte) = iter.next() {
+            // Convert the byte to a valid field symbol.
+            let field_symbol = FieldSymbol::try_from(byte)?;
+
+            // Go through the bytes to count how often it's repeated.
+            let mut field_length: u8 = 1;
+            while let Some(next_byte) = iter.peek() {
+                if *next_byte != byte {
+                    break;
+                }
+                field_length += 1;
+                iter.next();
+            }
+
+            if let Some(prev_field) = fields.last() {
+                in_order = in_order
+                    && prev_field.symbol.get_canonical_order() < field_symbol.get_canonical_order();
+            }
```

Comment:
Suggestion: If the field is out of order, `insert` it into the SmallVec.  An easy way to do this would be to walk backward from the end of the SmallVec until you find the position where you need to insert the field.  If the skeleton is already sorted, then you always need to check just one element, and you insert it at the end.

You could also use binary search for the correct position, which is what LiteMap does under the hood.

You would then be able to get rid of `SkeletonError::FieldsOutOfOrder` and the code to handle it elsewhere.

Reply:
Ah ok, I was trying to write code that would make it impossible to accidentally use unsorted fields. I guess it doesn't really matter that much if I accept unsorted fields here as bincode is the faster path. This change fits nicely with the ordering changes.